multicluster/controllers/multicluster/leader/stale_controller.go

```diff
@@ -0,0 +1,247 @@
+				"MemberClusterAnnounce", klog.KObj(&memberClusterAnnounce), "agreedPeriod", memberClusterAnnounceStaleTime)
+		} else {
+			klog.InfoS("Cleaning up stale MemberClusterAnnounce. The latest update time is not in RFC3339 format",
+				"MemberClusterAnnounce", klog.KObj(&memberClusterAnnounce))
+		}
+
+		if err := c.Client.Delete(ctx, &memberClusterAnnounce, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+			klog.ErrorS(err, "Failed to delete stale MemberClusterAnnounce", "MemberClusterAnnounce", klog.KObj(&memberClusterAnnounce))
+			return
+		}
+	}
+}
+
+func (c *LeaderStaleResCleanupController) RunPeriodically(stopCh <-chan struct{}) {
+	klog.InfoS("Starting LeaderStaleResCleanupController")
+	defer klog.InfoS("Shutting down LeaderStaleResCleanupController")
+
+	ctx, _ := wait.ContextForChannel(stopCh)
+	go wait.UntilWithContext(ctx, c.cleanUpStaleMemberClusterAnnounces, memberClusterAnnounceStaleTime/2)
+	<-stopCh
+}
+
+func (c *LeaderStaleResCleanupController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	memberAnnounce := &mcv1alpha1.MemberClusterAnnounce{}
+	err := c.Get(ctx, req.NamespacedName, memberAnnounce)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if err == nil && memberAnnounce.DeletionTimestamp.IsZero() {
+		// Ignore the event if it's not with non-zero DeletionTimestamp
+		return ctrl.Result{}, nil
+	}
+
+	// Clean up all corresponding ResourceExports when the member cluster's
+	// MemberClusterAnnounce is deleted.
+	clusterID := getClusterIDFromName(req.Name)
+	staleResExports, err := getResourceExportsByClusterIDFunc(c, ctx, clusterID)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get ResourceExports by ClusterID", "clusterID", clusterID)
+		return ctrl.Result{}, err
+	}
+	cleanUpSucceed := deleteResourceExports(ctx, c.Client, staleResExports)
+	if !cleanUpSucceed {
+		return ctrl.Result{}, fmt.Errorf("failed to clean up all stale ResourceExports for the member cluster %s, retry later", clusterID)
+	}
+
+	// When cleanup is done, remove the Finalizer of this MemberClusterAnnounce.
+	finalizer := fmt.Sprintf("%s/%s", MemberClusterAnnounceFinalizer, memberAnnounce.ClusterID)
+	memberAnnounce.Finalizers = common.RemoveStringFromSlice(memberAnnounce.Finalizers, finalizer)
+	if err := c.Update(context.TODO(), memberAnnounce); err != nil {
+		klog.ErrorS(err, "Failed to update MemberClusterAnnounce", "MemberClusterAnnounce", klog.KObj(memberAnnounce))
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (c *LeaderStaleResCleanupController) RunOnce(stopCh <-chan struct{}) {
+	// Try to clean up all stale ResourceExports before start in case
+	// there was a MemberClusterAnnounce deleted but the event was not handled
+	// properly if the Antrea Multi-cluster Controller was stopped.
+	retry.OnError(common.CleanUpRetry, func(err error) bool { return true },
```

Comment:
This can be removed, as with Finalizer a memberAnnounce cannot be deleted before Exports are cleaned up?

Reply:
Right, removed this.